globalsat_sport.cc

```diff
@@ -0,0 +1,742 @@
+	globalsat_probe_device ();
+}
+
+static void
+wr_init(const char *fname)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " wr_init()\n");
+	}
+	serial_init (fname);
+}
+
+
+static void
+rd_deinit(void)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " rd_deinit()\n");
+	}
+	serial_deinit();
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " rd_deinit() Done\n");
+	}
+}
+
+static void
+wr_deinit(void)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " wr_deinit()\n");
+	}
+	serial_deinit ();
+}
+
+static void track_read (void);
+
+
+static void
+waypoint_read(void)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME "   waypoint_read()\n");
+	}
+	//CommandGetTrackFileHeaders
+	globalsat_send_simple (CommandGetWaypoints);
+
+	uint8_t *in_payload;
+	int len;
+	uint8_t DeviceCommand;
+	in_payload = globalsat_read_package (&len, &DeviceCommand);
+	if ((len > 0) && (in_payload != NULL)) {
+		if (global_opts.debug_level > 1) {
+			printf ("Got package!!!\n");
+		}
+		//TODO figure out what device it is
+	}
+	if (in_payload) {
+		free (in_payload);
+	}
+	track_read ();
+}
+
+static void
+track_read(void)
+{
+	route_head *trk = NULL;
+	Waypoint *wpt = NULL;
+    gpsbabel::DateTime gpsDateTime;
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME "   track_read()\n");
+	}
+	//CommandGetTrackFileHeaders
+	globalsat_send_simple (CommandGetTrackFileHeaders);
+	if (global_opts.debug_level > 1) {
+		printf ("Sent...\n");
+	}
+
+	uint8_t *payload;
+	int length;
+	uint8_t DeviceCommand;
+	payload = globalsat_read_package (&length, &DeviceCommand);
+	if ((length > 0) && (payload != NULL)) {
+		if (global_opts.debug_level > 1) {
+			printf ("Got package!!! headers\n");
+		}
+		int number_headers;
+		number_headers = length / sizeof (gh_trainheader);	//29=packed sizeof(gh_trainheader)
+		if (global_opts.debug_level > 1) {
+			printf ("lenght=%d sizeof(gh_trainheader)=%d number_headers=%d\n", (int) length, (int) sizeof (gh_trainheader), (int) number_headers);
+		}
+
+		int i;
+		for (i = 0; i < number_headers; i++) {
+			int pos = i * sizeof (gh_trainheader);
+			gh_trainheader *header = (gh_trainheader *) & (payload[pos]);
+			if (global_opts.debug_level > 1) {
+				printf ("%02d-%02d-%02d ", header->dateStart.Year,header->dateStart.Month, header->dateStart.Day);
+				printf ("%02d:%02d:%02d ", header->timeStart.Hour,header->timeStart.Minute, header->timeStart.Second);
+				printf ("Total(points:%6d time:%6ds dist:%9dm) LapCnts:%5d ",be32toh (header->TotalPoint),be32toh (header->TotalTime) / 10,	be32toh (header->TotalDistance), be16toh (header->LapCnts));
+				printf ("Index/StartPt:%d ", be32toh (header->gh_ptrec.Index));
+				printf ("LapIndex/EndPt:%d ", be32toh (header->gh_laprec.LapIndex));
+				printf ("DataType:0x%x\n", header->DataType);
+			}
+		}
+
+		for (i = 0; i < number_headers; i++) {
+			int pos = i * sizeof (gh_trainheader);
+			gh_trainheader *header = (gh_trainheader *) & (payload[pos]);
```

Comment:
Will we get into trouble type-punning here if the gh_trainheader and payload have different alignment requirements?